### ISSUE TITLE
Read timeout to candidate cluster using HTTPS

### ISSUE THREAD

**Issue by hatdropper1977** [elastic/rally#391]
**Rally version** (get with `esrally --version`):
```bash
(rally)[centos@ip-10-0-0-48 ~]$ esrally --version
esrally 0.8.1
```

**Invoked command**:
```bash
(rally)[centos@ip-10-0-0-48 ~]$ esrally --track=pmc --revision=latest --distribution-version=6.1.1 --user-tag="intention:troubleshoot" --elasticsearch-plugins=x-pack:security --client-options="use_ssl:true,verify_certs:false,basic_auth_user:'rally',basic_auth_password:'rally-password'" --cluster-health=yellow --on-error=abort
```

**Configuration file (located in `~/.rally/rally.ini`))**:
```yaml
[system]
env.name = local

[node]
root.dir = /mnt/data1
src.root.dir = /mnt/data1/src

[source]
remote.repo.url = https://github.com/elastic/elasticsearch.git
elasticsearch.src.subdir = elasticsearch

[build]
gradle.bin = /usr/bin/gradle

[runtime]
java.home = /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.151-5.b12.el7_4.x86_64

[benchmarks]
local.dataset.cache = ${node:root.dir}/data

[reporting]
datastore.type = in-memory
datastore.host =
datastore.port =
datastore.secure =
datastore.user =
datastore.password =

[tracks]
default.url = https://github.com/elastic/rally-tracks

[teams]
default.url = https://github.com/elastic/rally-teams

[defaults]
preserve_benchmark_candidate = False

[distributions]
release.1.url = https://download.elasticsearch.org/elasticsearch/elasticsearch/elasticsearch-{{VERSION}}.tar.gz
release.2.url = https://download.elasticsearch.org/elasticsearch/release/org/elasticsearch/distribution/tar/elasticsearch/{{VERSION}}/elasticsearch-{{VERSION}}.tar.gz
release.url = https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-{{VERSION}}.tar.gz
release.cache = true
```

**JVM version**:
```bash
(rally)[centos@ip-10-0-0-48 ~]$ java -version
openjdk version "1.8.0_151"
OpenJDK Runtime Environment (build 1.8.0_151-b12)
OpenJDK 64-Bit Server VM (build 25.151-b12, mixed mode)
```

**OS version**:
```bash
(rally)[centos@ip-10-0-0-48 ~]$ uname -a
Linux ip-10-0-0-48.ec2.internal 3.10.0-693.5.2.el7.x86_64 #1 SMP Fri Oct 20 20:32:50 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Description of the problem including expected versus actual behavior**:
I attempted to run an in-memory race of PMC using the x-pack plugin.  In this race, I do not connect to an existing cluster, I have rally download a server.  I use Elasticsearch version 6.6.1  I used the correct flags as noted above in the command section (SSL Verify off).  The race fails because the rally service cannot connect to the HTTPS server.  I am running Rally via a Python virtual environment.  

**Steps to reproduce**:
Run the rally command above.

**Provide logs (if relevant)**:
```bash
Running index-append                                                           [ 11% done]
[ERROR] Cannot race. ('Error in load generator [3]', Request returned an error. Error type: transport, Description: HTTPSConnectionPool(host='127.0.0.1', port=39200): Read timed out. (read timeout=10) (HTTPSConnectionPool(host='127.0.0.1', port=39200): Read timed out. (read timeout=10)))
```

```python
2018-01-11 19:31:59,996 PID:21837 rally.mechanic INFO Child actor exited while engine is stopping or stopped: [ChildActorExited:ActorAddr-(T|:41352)]
2018-01-11 19:31:59,998 PID:21836 rally.racecontrol INFO BenchmarkActor#receiveMessage(msg = [<class 'thespian.actors.ChildActorExited'>] sender = [ActorAddr-(T|:33397)])
2018-01-11 19:31:59,998 PID:21836 rally.racecontrol INFO BenchmarkActor received unknown message [ChildActorExited:ActorAddr-(T|:33397)] (ignoring).
2018-01-11 19:32:00,2 PID:21834 root INFO ---- Actor System shutdown
2018-01-11 19:32:00,788 PID:21828 rally.main INFO Shutdown completed.
2018-01-11 19:32:00,788 PID:21828 root ERROR Cannot run subcommand [race].
Traceback (most recent call last):
  File "/home/centos/rally/lib/python3.4/site-packages/esrally/rally.py", line 530, in dispatch_sub_command
    race(cfg)
  File "/home/centos/rally/lib/python3.4/site-packages/esrally/rally.py", line 460, in race
    with_actor_system(lambda c: racecontrol.run(c), cfg)
  File "/home/centos/rally/lib/python3.4/site-packages/esrally/rally.py", line 480, in with_actor_system
    runnable(cfg)
  File "/home/centos/rally/lib/python3.4/site-packages/esrally/rally.py", line 460, in <lambda>
    with_actor_system(lambda c: racecontrol.run(c), cfg)
  File "/home/centos/rally/lib/python3.4/site-packages/esrally/racecontrol.py", line 382, in run
    raise e
  File "/home/centos/rally/lib/python3.4/site-packages/esrally/racecontrol.py", line 379, in run
    pipeline(cfg)
  File "/home/centos/rally/lib/python3.4/site-packages/esrally/racecontrol.py", line 63, in __call__
    self.target(cfg)
  File "/home/centos/rally/lib/python3.4/site-packages/esrally/racecontrol.py", line 319, in from_distribution
    return race(cfg, distribution=True)
  File "/home/centos/rally/lib/python3.4/site-packages/esrally/racecontrol.py", line 281, in race
    raise exceptions.RallyError(result.message, result.cause)
esrally.exceptions.RallyError: ('Error in load generator [3]', Request returned an error. Error type: transport, Description: HTTPSConnectionPool(host='127.0.0.1', port=39200): Read timed out. (read timeout=10) (HTTPSConnectionPool(host='127.0.0.1', port=39200): Read timed out. (read timeout=10)))
```
[rally_out_20180111T192948Z.log](https://github.com/elastic/rally/files/1623900/rally_out_20180111T192948Z.log)
[rally-actors.log](https://github.com/elastic/rally/files/1623901/rally-actors.log)
[rally-actor-messages.log](https://github.com/elastic/rally/files/1623902/rally-actor-messages.log)

**Comment by hatdropper1977** [comment]
It turns out that I needed to disable SE Linux.

**Comment by hatdropper1977** [comment]
Never mind.  It ran succesfully once and then I tried to re-run it and it failed again.

**Comment by hatdropper1977** [comment]
More information.

I used the following command:
```bash
esrally --track=pmc --revision=latest --distribution-version=6.1.1 --user-tag="intention:Rally-Downloads-Server-Baseline-X-Pack-rfr-gold" --elasticsearch-plugins=x-pack:security --client-options="use_ssl:true,verify_certs:false,basic_auth_user:'rally',basic_auth_password:'rally-password'" --cluster-health=yellow --on-error=abort --car="16gheap" 
```
That is, a car with 16g heap.  I set max file descriptors to 65536.

```bash
rally_out_20180111T221217Z.log:2018-01-11 22:12:33,284 PID:2022 rally.launcher INFO rally-node-0: [2018-01-11T22:12:33,283][WARN ][o.e.b.BootstrapChecks    ] [rally-node-0] max file descriptors [65535] for elasticsearch process is too low, increase to at least [65536]
```

I deleted documents.json.offset and the distributions, races, teams and tracks directories (I left the data).

The race runs for a bit (64% on index-append)and then it eventually breaks.

```bash
(rally)[centos@ip-10-0-0-48 ~]$ esrally --track=pmc --revision=latest --distribution-version=6.1.1 --user-tag="intention:Rally-Downloads-Server-Baseline-X-Pack-rfr-gold" --elasticsearch-plugins=x-pack:security --client-options="use_ssl:true,verify_certs:false,basic_auth_user:'rally',basic_auth_password:'rally-password'" --cluster-health=yellow --on-error=abort --car="16gheap" 

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[WARNING] --cluster-health is deprecated and will be removed in a future version of Rally.
[INFO] You did not provide an explicit timeout in the client options. Assuming default of 10 seconds.
[INFO] Writing logs to /home/centos/.rally/logs/rally_out_20180111T221217Z.log
[INFO] Preparing for race ...
[INFO] Downloading Elasticsearch 6.1.1 (27.1 MB total size)                         [100%]
[INFO] Rally will delete the benchmark candidate after the benchmark
[INFO] Racing on track [pmc], challenge [append-no-conflicts] and car ['16gheap'] with version [6.1.1].

[INFO] Preparing file offset table for [/mnt/data1/data/pmc/documents.json] ... [OK]
Running index-append                                                           [ 64% done]


[WARNING] Could not terminate all internal processes within timeout. Please check and force-terminate all Rally processes.
[ERROR] Cannot race. ('Error in load generator [1]', Request returned an error. Error type: transport, Description: HTTPSConnectionPool(host='127.0.0.1', port=39200): Read timed out. (read timeout=10) (HTTPSConnectionPool(host='127.0.0.1', port=39200): Read timed out. (read timeout=10)))

Getting further help:
*********************
* Check the log files in /home/centos/.rally/logs for errors.
* Read the documentation at https://esrally.readthedocs.io/en/0.8.1/
* Ask a question on the forum at https://discuss.elastic.co/c/elasticsearch/rally
* Raise an issue at https://github.com/elastic/rally/issues and include the log files in /home/centos/.rally/logs.

---------------------------------
[INFO] FAILURE (took 436 seconds)
```

The logs show that a _bulk action failed.

```bash
2018-01-11 22:19:03,550 PID:2880 rally.driver INFO LoadGenerator[5] is executing [index-append] (65.97% complete).
2018-01-11 22:19:03,551 PID:2797 rally.driver INFO Driver#receiveMessage(msg = [<class 'esrally.driver.driver.UpdateSamples'>] sender = [ActorAddr-(T|:33442)])
2018-01-11 22:19:03,551 PID:2797 rally.driver INFO Driver#receiveMessage(msg = [<class 'esrally.driver.driver.UpdateSamples'>] sender = [ActorAddr-(T|:44936)])
2018-01-11 22:19:03,580 PID:2797 rally.driver INFO Driver#receiveMessage(msg = [<class 'thespian.actors.WakeupMessage'>] sender = [ActorAddr-(T|:39243)])
2018-01-11 22:19:03,580 PID:2881 rally.driver INFO LoadGenerator[6]#receiveMessage(msg = [<class 'thespian.actors.WakeupMessage'>], sender = [ActorAddr-(T|:45989)])
2018-01-11 22:19:03,580 PID:2881 rally.driver INFO LoadGenerator[6] is executing (no samples).
2018-01-11 22:19:03,630 PID:2877 rally.driver INFO LoadGenerator[2]#receiveMessage(msg = [<class 'thespian.actors.WakeupMessage'>], sender = [ActorAddr-(T|:42471)])
2018-01-11 22:19:03,630 PID:2877 rally.driver INFO LoadGenerator[2] is executing (no samples).
2018-01-11 22:19:04,590 PID:2797 rally.driver INFO Driver#receiveMessage(msg = [<class 'thespian.actors.WakeupMessage'>] sender = [ActorAddr-(T|:39243)])
2018-01-11 22:19:05,600 PID:2797 rally.driver INFO Driver#receiveMessage(msg = [<class 'thespian.actors.WakeupMessage'>] sender = [ActorAddr-(T|:39243)])
2018-01-11 22:19:06,600 PID:2797 rally.driver INFO Driver#receiveMessage(msg = [<class 'thespian.actors.WakeupMessage'>] sender = [ActorAddr-(T|:39243)])
2018-01-11 22:19:07,259 PID:2022 rally.mechanic INFO Flushing system metrics store on host [127.0.0.1].
2018-01-11 22:19:07,274 PID:2022 rally.metrics INFO Successfully added 30 metrics documents for invocation=[20180111T221217Z], track=[pmc], challenge=[append-no-conflicts], car=[['16gheap']].
2018-01-11 22:19:07,610 PID:2797 rally.driver INFO Driver#receiveMessage(msg = [<class 'thespian.actors.WakeupMessage'>] sender = [ActorAddr-(T|:39243)])
2018-01-11 22:19:08,280 PID:2876 elasticsearch WARNING POST https://127.0.0.1:39200/_bulk [status:N/A request:10.078s]
Traceback (most recent call last):
  File "/home/centos/rally/lib/python3.4/site-packages/urllib3/connectionpool.py", line 386, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/home/centos/rally/lib/python3.4/site-packages/urllib3/connectionpool.py", line 382, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib64/python3.4/http/client.py", line 1227, in getresponse
    response.begin()
  File "/usr/lib64/python3.4/http/client.py", line 386, in begin
    version, status, reason = self._read_status()
  File "/usr/lib64/python3.4/http/client.py", line 348, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib64/python3.4/socket.py", line 378, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib64/python3.4/ssl.py", line 748, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib64/python3.4/ssl.py", line 620, in read
    v = self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out
```

**Comment by danielmitterdorfer** [comment]
Hmm, this looks like a perfectly valid condition to me. It's just that you encounter a read timeout. You can increase it by adding the `timeout` parameter to your client options, e.g.:

```
--client-options="timeout:60,use_ssl:true,verify_certs:false,basic_auth_user:'rally',basic_auth_password:'rally-password'"
```

will increase the timeout to 60 seconds.

Can you please retry with the modified settings?

**Comment by hatdropper1977** [comment]
Yes - that helps.  

Since I just had the results store in memory, it didn't occur to me to update the _Rally server's_ OS with the [important system configurations](https://www.elastic.co/guide/en/elasticsearch/reference/master/system-config.html).

Once I set **nproc** and **nofile** on the _Rally server OS_, I was able to run a 32gheap car with the default time out.

Is this worth putting into the documentation?  A gentler reminder to people who run out of the box that they still should set the local OS kernel parameters?  I can add a 'note' to the [prerequisites](http://esrally.readthedocs.io/en/stable/install.html) documentation.

**Comment by danielmitterdorfer** [comment]
> Is this worth putting into the documentation? A gentler reminder to people who run out of the box that they still should set the local OS kernel parameters?

Yes, I think that would make sense.

> I can add a 'note' to the prerequisites documentation.

Thanks. This seems to be a good place at the moment. Sooner or later I think it would be good to have a dedicated page in the docs with benchmarking best practices and we can then later move this there. But as I said, currently the prerequisites section makes sense to me.

**Comment by danielmitterdorfer** [comment]
Closing as the docs change is now addressed by #397.
